apps/common/lib/lexical/ast/env.ex

```diff
@@ -415,12 +423,52 @@ defmodule Lexical.Ast.Env do
     end)
   end
 
+  @type_keys [:type, :typep, :opaque]
   defp ancestor_is_type?(env) do
     env.document
     |> Ast.cursor_path(env.position)
     |> Enum.any?(fn
-      {:type, _, _} -> true
-      _ -> false
+      {:@, metadata, [{type_key, _, _}]} when type_key in @type_keys ->
+        # single line type
+        cursor_in_range?(env, metadata)
+
+      {:__block__, _, [{:@, metadata, [{type_key, _, _}]}, _]}
+      when type_key in @type_keys ->
+        # multi-line type
+        cursor_in_range?(env, metadata)
+
+      _ ->
+        false
+    end)
+  end
+
+  defp ancestor_is_spec?(env) do
+    env.document
+    |> Ast.cursor_path(env.position)
+    |> Enum.any?(fn
+      {:@, metadata, [{:spec, _, _}]} ->
+        # single line spec
+        cursor_in_range?(env, metadata)
+
+      {:__block__, _, [{:@, metadata, [{:spec, _, _}]}, _]} ->
+        # multi-line spec
+        cursor_in_range?(env, metadata)
+
+      _ ->
+        false
     end)
   end
+
+  defp cursor_in_range?(env, metadata) do
```

Comment:
I would like to know the difference between this function and `Lexical.Ast .contains_position`?

Reply:
I was unaware of that function. I'll try it out and switch.

Reply:
actually scratch that. The difference is that `Ast.contains_position?` uses `Sourceror.get_range`, which examines a block of ast and tells you the extent of that.  This function takes a single expression and uses the extent of it in the metadata to figure out the range.